### ISSUE TITLE
[bug|windows] Infinite exit loop after setting the Monitor Trap Flag (MTF)

### ISSUE THREAD

**Issue by no-realm** [Randshot/ept_split#1]
At the moment, there is an issue, where, after setting the `MTF`, the `RIP` ends up in some kernel function (in `ntoskrnl.exe`).
This has the side effect, that the actual instruction, that we wanted to single step, not being executed, resulting in an infinite loop between `EPT Violation` and `MTF` exits.
Note that the above is only the case if *thrashing* was detected.

I haven't managed to find a solution yet, but a user on the **OSR Forum** seems to have a similar issue caused by interrupts.
[Semantics of Monitor Trap Flag (VT-x)](https://community.osr.com/discussion/285021)

**Comment by no-realm** [comment]
Fixed.

**Comment by rianquinn** [comment]
Did you ever fix this? We are seeing something similar with RIP being somewhere else and you end up with an infinite loop. Not really sure of the right way to fix this.

**Comment by rianquinn** [comment]
@Randshot I found this:
https://patchwork.kernel.org/patch/9715791/

They talk about how MTF is a mess. I am basically seeing exactly the same thing. When you enter back into the guest you end up in the kernel (I am assuming that an interrupt is being executed). I am not really sure why it happens sooooo reliably, but it does. If I leave MTF enabled, I can watch the kernel execute code and then execute the interrupt handler again and it just ends up in yet another infinite loop. I am assuming this is because outputting to serial is so slow that the kernel cannot service interrupts fast enough. 

So long story short is, I am not sure that MTF is all that useful. Either way, I would love to know how you fixed this on your end. I have an idea, but its not simple.

**Comment by no-realm** [comment]
@rianquinn 
I managed to find a passable solution for my (this) case.
First I reduced the amount of logging instructions on the "hot" path, and only output to serial if something unexpected happened.
The second thing I did, is to NOT flip between pages (change the physical address), but instead change the EPTP (pointer). This reduces the "thrashing"/exits significantly.

I don't know if that is a viable solution for you though.
A proper fix for this would also be nice, but I personally haven't found one.

**Comment by rianquinn** [comment]
Yeah, I would say if you don't care about security, reducing the likelihood of this ever happening is an easy way to solve the issue. 

For our hook example, I will likely be removing the use of MTF as I think it should be avoided as an example since it is so broken for a VMM. A better option would be what @tklengyel does with DRAKVUF which is to use the #BP and insert an int 0x03 into a shadow page. Another option would be to just modify a shadow page (which is what I will likely do for the example), but that only works for hooking specifically, and not passing through an instruction. 

In the case above, its a real problem because trying to pass-through an instruction like ENCLS is really hard to do without the guest being able to fool you in some way, but emulating this instruction in the Host is also really hard as the ENCLS instruction relies on guest specific state. 

Seems like there could be an argument to be made to make an improvement to VMX that would allow you to single-step a specific instruction in memory at a specific RIP. That would make MTF viable.

**Comment by tklengyel** [comment]
@rianquinn I still use MTF to switch between the shadow-page with the injected INT3 and the normal view exactly the same way @Randshot solved his issue: switching EPTPs.

**Comment by rianquinn** [comment]
Then technically, you should have the same issue. That is, if an interrupt
fires you would end up with MTF inside the kernel.

Or am I missing something?

**Comment by tklengyel** [comment]
Possibly, but from my end it doesn't really matter. At the MTF exit the EPT view is switched back to shadow page, so after the interrupt is handled the OS would just trap again when it resumes execution where it left off. This may appear as the trap happening 2x in the logs but otherwise it's not a problem.

**Comment by no-realm** [comment]
Yeah, I also am using MTF for cases where an instruction tries to read memory on the same page (like jump tables).
